packages/common/src/audius-query/createApi.ts

```diff
@@ -553,6 +557,130 @@ const buildEndpointHooks = <
     }
   }
 
+  const usePaginatedQuery = <Data extends [], ArgsType extends Args>(
+    baseArgs: Omit<ArgsType, 'limit' | 'offset'>,
+    options: PaginatedQueryHookOptions
+  ): PaginatedQueryHookResults<Data> => {
+    const [loadingMore, setLoadingMore] = useState(false)
+    const dispatch = useDispatch()
+    const { pageSize = 5, singlePageData, ...queryHookOptions } = options
+    const [forceLoad, setForceLoad] = useState(false)
+    const [page, setPage] = useState(0)
+    const [status, setStatus] = useState<Status>(Status.IDLE)
+
+    // A soft reset resets our current page and loading state, but the data is still held in store (which means cache hits will still return as expected)
+    // This is triggered below when the args or page size changes
+    const softReset = useCallback(() => {
+      setPage(0)
+      setLoadingMore(false)
+    }, [])
+
+    // Includes softReset above BUT also hard resets cached data in store; also resets loading status
+    // Can be used to hard refresh all data and reset the page
+    const hardReset = useCallback(() => {
+      softReset()
+      setStatus(Status.LOADING)
+      // @ts-ignore - Unclear whats wrong with they type here
+      dispatch(actions[`reset${capitalize(endpointName)}`]())
+    }, [dispatch, softReset])
+
+    // Only need to soft reset when args or pagesize changes - leaves all cached data intact
+    useCustomCompareEffect(softReset, [baseArgs], isEqual)
+
+    const args = {
+      ...baseArgs,
+      limit: pageSize,
+      offset: page * pageSize
+    } as ArgsType & PaginatedQueryArgs
+
+    const allPageData = useSelector((state: CommonState) => {
+      // If in single page mode we only return the current page of data (no aggregation)
+      if (singlePageData) {
+        return state?.api?.[reducerPath]?.[endpointName]?.[
+          getKeyFromFetchArgs(args)
+        ]?.normalizedData
+      }
+
+      // By default, select data for every page up to the current page and aggregate the results
+      const pageAccumulator = []
+
+      for (let i = 0; i <= page; i++) {
+        const key = getKeyFromFetchArgs({
+          ...args,
+          offset: i * pageSize
+        })
+        const normalizedPageData =
+          state?.api?.[reducerPath]?.[endpointName]?.[key]?.normalizedData
+        if (normalizedPageData) {
+          pageAccumulator.push(...normalizedPageData)
+        }
+      }
+      return pageAccumulator
+    })
+
+    // Since this is a paginated query we know that the results will be in an array
+    // Call the useQuery hook to trigger cache checks and/or fetches
+    const result = useQuery(args, queryHookOptions) as QueryHookResults<Data[]>
```

Comment:
we could change the selector to be able to hit the legacy cache, or maybe write to audius-query cache on cache hit inside `useQueryState`

Reply:
I dont think I'm following you? This didn't really change much from `useAllPaginatedQuery`, previously this was just being passed in but now we can just use it since it's in scope already